### ISSUE TITLE
as_scalar error in max_pooling based on number of maxIterations for optimizer.

### ISSUE THREAD

**Issue by saraansh1999** [mlpack/mlpack#2252]
* version of mlpack: 3.2.3 (output of `cout << MLPACK_VERSION_MAJOR << 
MLPACK_VERSION_MINOR << 
MLPACK_VERSION_PATCH;` )
 * operating system: Ubuntu 16.04
 * compiler: g++ (gcc version 5.4.0 20160609)
 * version of dependencies (Boost/Armadillo): 
Boost version: 1.58.0
ARMA version: 9.850.1 (Pyrocumulus Wrath)

Steps to reproduce:
[https://drive.google.com/drive/folders/1bHwfqIQcidgX8sOiwoMFdQWkWA9G4DnB?usp=sharing](url)
The code and the data files are present in this link.

Expected Behavior: The model should start training.

Actual Behavior:

Error:
`error: as_scalar(): expression doesn't evaluate to exactly one element terminate called after throwing an instance of 'std::logic_error' what(): as_scalar(): expression doesn't evaluate to exactly one element Aborted (core dumped)`

I was able to trace down the error to the line:
[https://github.com/mlpack/mlpack/blob/225ec4e951a5333aa6c84db1490c76e9ca873977/src/mlpack/methods/ann/layer/max_pooling.hpp#L36](url)

I don't know why this is happening.
Been stuck for a while now.
Might be something stupid.
Any help is appreciated.
TIA

**Comment by saraansh1999** [comment]
Pls tell if the problem is reproducible.
TIA

**Comment by zoq** [comment]
Before I take a closer look into the issue, is the model and data from the models repo?

**Comment by saraansh1999** [comment]
The code is almost the same as this
https://github.com/mlpack/models/blob/master/Kaggle/DigitRecognizerCNN/src/DigitRecognizerCNN.cpp
As suggested by a member on gitter

**Comment by zoq** [comment]
Did you test the original code? The reason why I ask is, that such information help a lot to figure out what went wrong as fast a possible.

**Comment by saraansh1999** [comment]
I did. 
Just to be sure i'll test it once more.

**Comment by zoq** [comment]
Thanks, also I guess it fails with the same error?

**Comment by saraansh1999** [comment]
Yup.
This is the gdb traceback to the maxpooling layer.
`
Starting program: /home/saraansh/gsoc2020try/mlpack_tutorials/a.out 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff276e700 (LWP 11452)]
[New Thread 0x7ffff1f6d700 (LWP 11453)]
[New Thread 0x7fffef76c700 (LWP 11454)]
[New Thread 0x7fffeaf6b700 (LWP 11455)]
[New Thread 0x7fffe876a700 (LWP 11456)]
[New Thread 0x7fffe5f69700 (LWP 11457)]
[New Thread 0x7fffe3768700 (LWP 11458)]
Reading data ...
Training ...
784 37800
1 37800

error: as_scalar(): expression doesn't evaluate to exactly one element
terminate called after throwing an instance of 'std::logic_error'
  what():  as_scalar(): expression doesn't evaluate to exactly one element

Thread 1 "a.out" received signal SIGABRT, Aborted.
0x00007ffff6915428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) up
#1  0x00007ffff691702a in __GI_abort () at abort.c:89
89	abort.c: No such file or directory.
(gdb) up
#2  0x00007ffff747a84d in __gnu_cxx::__verbose_terminate_handler() ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) up
#3  0x00007ffff74786b6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) up
#4  0x00007ffff7478701 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) up
#5  0x00007ffff7478919 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) up
#6  0x0000000000407a19 in void arma::arma_stop_logic_error<char const*>(char const* const&) ()
(gdb) up
#7  0x00000000004119e6 in void arma::arma_check<char [64]>(bool, char const (&) [64]) ()
(gdb) up
#8  0x00000000005216c3 in arma::mtOp<unsigned long long, arma::Mat<double>, arma::op_index_max>::elem_type arma::as_scalar<arma::mtOp<unsigned long long, arma::Mat<double>, arma::op_index_max> >(arma::Base<arma::mtOp<unsigned long long, arma::Mat<double>, arma::op_index_max>::elem_type, arma::mtOp<unsigned long long, arma::Mat<double>, arma::op_index_max> > const&) ()
(gdb) up
#9  0x00000000004fe020 in unsigned long mlpack::ann::MaxPoolingRule::Pooling<arma::Mat<double> >(arma::Mat<double> const&) ()

`

Once we are able to figure out the problem and if it needs some change in the code I would love to take it up.
Thanks

**Comment by rcurtin** [comment]
So, there's a `bug report` issue template that asks for all manner of things including, e.g., the version of Armadillo and dependencies.  Is there a reason that you didn't use that?  Because that information would be really helpful at this point. :)

**Comment by saraansh1999** [comment]
I wasnt sure wether this was a bug in the codebase or a fault on my part. I'll update it.

**Comment by zoq** [comment]
Thanks!

**Comment by kartikdutt18** [comment]
Hi @saraansh1999, I am not able to reproduce this on my system. Here are some the results of the results from DigitRecognizer and DigitRecognizerCNN

```
1 - accuracy: train = 95.2963%, valid = 94.2143%
2 - accuracy: train = 97.2778%, valid = 96.2143%
3 - accuracy: train = 97.8148%, valid = 96.2381%
4 - accuracy: train = 97.5397%, valid = 95.7143%
5 - accuracy: train = 98.3942%, valid = 96.5%
6 - accuracy: train = 98.5132%, valid = 96.7381%

```
and 

```
Epoch 0:        Training Accuracy = 49.6376%,   Validation Accuracy = 50.4048%
Epoch 1:        Training Accuracy = 66.1243%,   Validation Accuracy = 65.8571%
Epoch 2:        Training Accuracy = 73.6878%,   Validation Accuracy = 73.3333%
Epoch 3:        Training Accuracy = 77.5952%,   Validation Accuracy = 77.5%
Epoch 4:        Training Accuracy = 80.6614%,   Validation Accuracy = 80.5238%
```

I think these are acceptable epoch accuracies. I will try running your code also and get back to you.
I will also run some tests on pooling layers just in case.

**Comment by saraansh1999** [comment]
Hey @kartikdutt18. Thanks for the effort. Following up on your comment I tried running both DigitRecognizer and DigitRecognizerCNN. The linear model seems to be working(i.e. DigitRecognizer), but the DigitRecognozerCNN still seems to be giving me the same error at the same place. 

I also tried making the value of MAX_ITERATIONS_PER_CYCLE same as trainX.n_cols(similar to the DigitRecognizer code) but no luck.

The changes I have made to run the DigitRecognizerCNN code, just to give completer info:
1. Download just the DigitRecognizerCNN and utils files from the repo.
2. Put the both in the same folder and accordingly change the include statement.
3. Change the paths to the dataset (the data is downloaded from the repo only).
4. Run using the command
` g++ DigitRecognizerCNN.cpp -fopenmp -larmadillo -lmlpack -std=c++11; ./a.out`

**Comment by kartikdutt18** [comment]
Hmmm, I couldn't get the code to fail but got very low accuracy (around 8% and it stayed there for a while).
`Training Accuracy = 9.39947%,   Validation Accuracy = 9.7166%`

Solution that worked for me:
The Following are the changes I made:

1. Increase iterations per cycle

`constexpr int ITERATIONS_PER_CYCLE = 10000;`

2. Reset Optimizer Every cycle

`optimizer.ResetPolicy()=false;`

3. Cleaned `cout` statement a bit. (irrelevant).

This fixes everything for me. The results I got:
```
Training Accuracy = 49.6376%
Validation Accuracy = 50.4048%

Training Accuracy = 66.1243%
Validation Accuracy = 65.8571%
```

The code I took is the one you have provided the link for in the description.

**Comment by kartikdutt18** [comment]
Could you also try running the model in model repo using
```
mkdir build && cd build && cmake ../
./bin/DigitRecognizerCNN
```
incase you haven't already.

**Comment by saraansh1999** [comment]
Same result  :(

**Comment by kartikdutt18** [comment]
Did you try both methods?

**Comment by saraansh1999** [comment]
The first option doesnt work as the code is not even passing the Train call.
Secondly i built the models repo and executed the command you mentioned. Same result.

**Comment by kartikdutt18** [comment]
Hmm, I will try it on Ubuntu then because I am not facing any issue on my local devices (OSX).

**Comment by saraansh1999** [comment]
Any updates?

**Comment by kartikdutt18** [comment]
@saraansh1999, I have my embedded systems exam tomorrow. So I am only be able to work on reviews / very minor changes. I will get back to you the day after unless someone beats me to it. I hope that's okay.

**Comment by kartikdutt18** [comment]
Hi  @saraansh1999, I tried it on linux based system(PopOS) and on MacOS. DigitRecognizerCNN works fine for me on both machines. I will run some tests to see why it fails on your machine, I'll try to get back to you as soon as I can.

**Comment by kartikdutt18** [comment]
Hey @saraansh1999, I have opened a PR that might be related to your issue. Not sure though but I think it is something to look into. Thanks.

**Comment by saraansh1999** [comment]
Please can I contribute somehow to this solution. I want to start with something small and get the feel of writing code in this codebase.

**Comment by kartikdutt18** [comment]
Hey I think this should be solved in #2265, Could you try it and see if it solved your issue?
Thanks.
